aQute.libg/src/aQute/lib/formatter/Formatters.java

```diff
@@ -0,0 +1,150 @@
+package aQute.lib.formatter;
+
+import java.time.ZonedDateTime;
+import java.util.Formatter;
+import java.util.Locale;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import aQute.lib.date.Dates;
+
+public class Formatters {
+	private final static Pattern PRINTF_P = Pattern.compile(
+		"%(?:(?<index>\\d+)\\$)?(?<previous><)?(-|\\+|0|\\(|,|\\^|#| )*(\\d*)?(?:\\.(\\d+))?(?<conversion>a|A|b|B|c|C|d|e|E|f|g|G|h|H|n|o|s|S|x|X|(?:[tT][HIklMSLNpzZsQBbhAaCYyjmdeRTrDFc])|%)");
```

Comment:
Argument index can be `<` as an alternate to `1$`. The regex implies you can say both, e.g. `1$<`, which is not the case.

`(?:(?<index>\\d+)\\$)?(?<previous><)?`

I think this should be 

`(?:(?:(?<index>\\d+)\\$)|(?<previous><))?`

```
Argument Index

The argument index is a decimal integer indicating the position of the argument in the argument list. The first argument is referenced by "1$", the second by "2$", etc.

Another way to reference arguments by position is to use the '<' ('\u003c') flag, which causes the argument for the previous format specifier to be re-used. For example, the following two statements would produce identical strings:

       Calendar c = ...;
       String s1 = String.format("Duke's Birthday: %1$tm %1$te,%1$tY", c);

       String s2 = String.format("Duke's Birthday: %1$tm %<te,%<tY", c);
 ```

Reply:
That will get caught on the actual formatting and it makes the regex even more complicated than it already is.

Reply:
a) it does not matter since the error will be found by the real formatter, we do not need fidelity. We just need to be able to parse anything that is legal for the formatter but it is not a problem if we parse wrong strings.
b) you do not need a group 2
c) you do not need ?: when you use named groups